language/move-prover/spec-lang/src/translate.rs

```diff
@@ -946,14 +946,43 @@ impl<'env, 'translator> ModuleTranslator<'env, 'translator> {
             _ => (None, None),
         };
         if let Some(name) = name {
+            let kind = match kind {
+                PA::SpecConditionKind::Assert => ConditionKind::Assert,
+                PA::SpecConditionKind::Assume => ConditionKind::Assume,
+                PA::SpecConditionKind::Decreases => ConditionKind::Decreases,
+                PA::SpecConditionKind::Ensures => ConditionKind::Ensures,
+                PA::SpecConditionKind::Requires => ConditionKind::Requires,
+                PA::SpecConditionKind::AbortsIf => ConditionKind::AbortsIf,
+            };
+            if kind == ConditionKind::Decreases {
+                self.parent
+                    .error(loc, "decreases specification not allowed currently");
+                return;
+            }
+            match code_offset {
+                None => {
+                    if !kind.on_decl() {
+                        self.parent
+                            .error(loc, "item only allowed inside function body");
+                        return;
+                    }
+                }
+                Some(_) => {
+                    if kind.on_decl() {
+                        self.parent
+                            .error(loc, "item only allowed on function declaration");
+                        return;
+                    }
+                }
+            }
             // We ensured that SpecBlockContext only contains resolvable names.
             let entry = &self
                 .parent
                 .fun_table
                 .get(name)
                 .expect("invalid spec block context")
                 .clone();
-            let is_precond = kind == &PA::SpecConditionKind::Requires;
+            let is_precond = kind == ConditionKind::Requires;
             let mut et = ExpTranslator::new(
                 self,
                 if is_precond {
```

Comment:
I believe old(_) is not allowed in on_impl conditions, right? Then this should be also OldExpStatus::NotSupported. Also, a little below in the code, if !is_precond we add the symbols result_1, result_2, ... which aren't available in impl asserts.

BTW, I generally would split this in two different functions (def_ana_condition_on_decl and def_ana_condition_on_impl). These have many differences, specifically once you include the locals delivered by Todds analyzes. Folding this all into one function lets this have too many cases and gets to complex.

Reply:
I thought a bit more about old(_) on impl conditions and I think we have to allow that to enable proving postconditions that refer to old(_).  Imagine having to prove such a postcondition for a function with a loop whose loop invariant would also then have to refer to old(...).  Boogie allows old(_) in asserts and assumes also.

Reply:
I am fixing the other issues.

Reply:
But what does old mean for a local variable? Or a spec variable? It's clear for parameters, but not in my mind for locals. We need to have a clear definition for this on Move level, and preserve its semantics when translating to Boogie.

Reply:
Right.  We have to be careful here.  For variables which can occur unambiguously inside old(_) in a postcondition, we want to provide the same meaning.  These would include spec variables (which are essentially like mutable global variables) and input parameters.  For other variables, such as local variables, I would prefer to disallow occurrence inside old(_) because otherwise the user will be confused about the semantics.